### DISCUSSION TITLE
Add "Referer" property to HttpTileSource

### DISCUSSION THREAD

**Discussion by gpsaliola** [Mapsui/Mapsui#2117] [Ideas]
I've seen that HttpTileSource has the property "userAgent" to specify the relative http header.
It would be great if there was also the "Referer" property.
In this way developers are not forced to implement tileFetcher to add this http header.
Some maps require this http header () such as Mapy maps https://mapserver.mapy.cz/turist-m/{z}-{x}-{y}  (Referer: "https://en.mapy.cz/").
Thanks.

**Comment by pauldendulk** [comment]
The userAgent is an optional parameter to the HttpTileSource. It is used in the OpenStreetMap class of Mapsui:

```csharp
    public static TileLayer CreateTileLayer(string? userAgent = null)
    {
        userAgent ??= $"user-agent-of-{Path.GetFileNameWithoutExtension(System.AppDomain.CurrentDomain.FriendlyName)}";

        return new TileLayer(CreateTileSource(userAgent)) { Name = "OpenStreetMap" };
    }

    private static HttpTileSource CreateTileSource(string userAgent)
    {
        return new HttpTileSource(new GlobalSphericalMercator(),
            "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
            new[] { "a", "b", "c" }, name: "OpenStreetMap",
            attribution: OpenStreetMapAttribution, userAgent: userAgent, persistentCache: DefaultCache);
    }
```

**Comment by gpsaliola** [comment]
It would also be useful to implement the Referer property required by some services (e.g. Mapy maps).
This would avoid implementing tileFetcher.
It's only a suggestion.
Thanks.

**Comment by gpsaliola** [comment]
In HTTP headers, there is not only the User-Agent property but many others. 
A widely used property is Referer and from what I understand it cannot be used with the code you showed me. 
If I understand correctly, it is only good for User-Agent. I proposed to also add Referer property used for example in Mapy maps.

**Reply by pauldendulk** [reply]
I am sorry, I misread your question!

I think will think about a solution. One reason not to add public setters is that it would become mutable, and we would need to take that into account in the logic. Adding every possible parameter in the in the constructor is not a great idea. A possible improvement would be to turn it into ```init``` properties.

**Comment by gpsaliola** [comment]
Good idea, although for web services purposes only User-Agent and Referer are important.
Without setting Referer Mapy maps cannot be displayed.

**Reply by pauldendulk** [reply]
Do you have an example of a public map service which needs this, so I can test this?

**Comment by gpsaliola** [comment]
These are map of Mapy serie.

https://mapserver.mapy.cz/base-m/{z}-{x}-{y}
https://mapserver.mapy.cz/turist-m/{z}-{x}-{y}
https://mapserver.mapy.cz/ophoto1012-m/{z}-{x}-{y}
https://mapserver.mapy.cz/winter-m/{z}-{x}-{y}

In particular Mapy Hiking is very interesting for hiking.

![Immagine 2023-09-16 141440](https://github.com/Mapsui/Mapsui/assets/6804212/584a09da-f611-4c65-8164-1e6c0841cf65)

**Comment by pauldendulk** [comment]
Thanks. I got this to work in the upcoming BruTile version by setting the referer. Here is samples code. Note, this will only work in the next release of BruTile:

```csharp
var mapy = new HttpTileSource(new GlobalSphericalMercator(Math.Max(0, minZoomLevel), Math.Min(18, maxZoomLevel)),
    "https://mapserver.mapy.cz/ophoto1012-m/{z}-{x}-{y}",
    name: source.ToString(),
    persistentCache: persistentCache, tileFetcher: tileFetcher,
    attribution: OpenStreetMapAttribution, userAgent: userAgent, referer: "https://en.mapy.cz/"),
```

**Comment by gpsaliola** [comment]
Magnificent!
This implementation will avoid implementing tileFetcher in all these cases.
Thank you very much.

**Comment by pauldendulk** [comment]
Done with this PR https://github.com/Mapsui/Mapsui/pull/2181
